### ISSUE TITLE
IDE0059 is not always reported

### ISSUE THREAD

**Issue by DaniilPalii** [dotnet/roslyn#52292]
**Version Used**: 

- .NET 5.0
- Roslynator.Analyzers 3.1.0

**Code to Reproduce**:
```csharp
    public class TestUnusedVariable
    {
        public void TestUnusedVariableInTrySection()
        {
            try
            {
                var unusedVariable = new object();
            }
            catch { }
        }

        public void TestUnusedVariableWhenUsingSelect()
        {
            var unusedVariable = new object();

            var selectResult = Enumerable.Empty<object>().Select(x => x);
        }

        public void TestUnusedVariableWhenUsingWhere()
        {
            var unusedVariable = new object();

            var whereResult = Enumerable.Empty<object>().Where(_ => true);
        }
    }
```

**Expected Behavior**:
Warning "IDE0059: Remove unnecessary value assignment" should be reported for each `unusedVariable` in the code above.

**Actual Behavior**:
Warning "IDE0059" is not reported.

**Comment by DaniilPalii** [comment]
I should add that the warning is reported in those cases

```csharp
    public class TestUnusedVariable
    {
        public void SimpleTestUnusedVariable()
        {
            var unusedVariable = new object();
        }

        public static void TestUnusedVariableWhenNotUsingLinq()
        {
            var unusedVariable = new object();

            var selectResult = Enumerable.Empty<object>();
        }
    }

**Comment by LukasGelke** [comment]
this also affects `foreach` loops. I'm "adding" to this instead of opening a new Issue, since it seems in both examples, the use of some Linq-Stuff "disables" the Analyzer.
We had a bug in production with `foreach (... in Enumerable.Chunk())` which should be caught by the analyzer.
```cs
// net6.0

public class Class2
{
  private void Do1(object[] objects)
  {
    foreach (object o in objects) { } // IDE0059 - OK
  }

  private void Do2(object[] objects)
  {
    foreach (object o in objects) { } // IDE0059 - OK
    foreach (object o in objects) { } // IDE0059 - OK
    string text = "abc";
  }

  private void Do3(object[] objects)
  {
    foreach (object o in objects) { } // MISSING
    foreach (object o in objects) { } // MISSING
    object[] others = [new()];
  }

  private void Do4(object[] objects)
  {
    foreach (object o in objects) { } // MISSING
    List<int> others = [new()];
  }

  private void Do5(object[] objects)
  {
    foreach (object o in objects) { } // MISSING
    foreach (object o in objects.Where(x => x.GetHashCode() != 0)) { } // MISSING??
    // not sure because of linq and sidefects, but it would still assign `o`; and this is like our bug
  }

  private void Do6((object, object)[] objects)
  {
    foreach ((object o, object b) in objects) { } // IDE0059 *2 - OK
  }

  private void Do7((object, object)[] objects)
  {
    foreach ((object o, object b) in objects) { } // MISSING *2
    object[] others = [new()];
  }
}
```

and a simplified version of the bug in question:
```cs
public class Class3
{ 
  private void ProdBug(MyDto[] objects)
  {
    foreach (MyDto[] o in objects.Chunk(5)) // MISSING??
    {
      DtoWorker(objects.Select(x => x.ammount));
    }
  }

  private void DtoWorker(IEnumerable<int> chunk)
  {
    // imagine a hard limit of 5
    _ = chunk;
  }

  private readonly record struct MyDto(int ammount);
}
```

**Comment by davidkarlsson** [comment]
@LukasGelke Nothing about foreach loops specifically are causing IDE0059 to not be reported for me on .NET 9 at least but having unused vars inside of a try catch block or in the same method as where a LINQ lambda is used does. 

This pretty much forces me to use Rider instead so that I can see unused vars reliably using resharper_unused_variable_highlighting instead so it's a shame it doesn't seem to be higher prioritized.

**Comment by CyrusNajmabadi** [comment]
>  it's a shame it doesn't seem to be higher prioritized.

We'd def accept community PRs for issues you think are higher priority for your needs :)

**Comment by LukasGelke** [comment]
@davidkarlsson 
I just re-tested, and using `Visual Studio Version 17.14.8 Preview 1.0` and `<TargetFramework>net9.0</TargetFramework>` I'm still missing Diagnostics in my `Class2.Do5` and `Class3` examples.
Previously missing cases `Do3`, `Do4` and `Do7` now get reported as expected
